text/0000-promise.md

```diff
@@ -0,0 +1,260 @@
+- Feature Name: `JsPromise`
+- Start Date: 2020-09-17
+- RFC PR: (leave this empty)
+- Neon Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Provide an API for creating, resolving and rejecting JavaScript Promises.
+
+```rust
+fn return_a_promise(cx: FunctionContext) -> JsResult<JsPromise> {
+    let (promise, deferred) = cx.promise();
+    let msg = cx.string("Hello, World!");
+
+    deferred.resolve(&cx, msg);
+
+    Ok(promise)
+}
+```
+
+# Motivation
+[motivation]: #motivation
+
+JavaScript Promise support has been a [long requested feature](https://github.com/neon-bindings/neon/issues/73). Promises are desirable for several reasons:
+
+* Considered best practices in idiomatic JavaScript
+* Enable `await` syntax for asynchronous operations
+* More easily map to asynchronous operations in native code
+
+Additionally, they can be combined with the [`EventQueue`](https://github.com/neon-bindings/rfcs/pull/32) API for very simple asynchronous threaded usage.
+
+```rust
+fn real_threads(cx: FunctionContext) -> JsResult<JsPromise> {
+    let queue = cx.queue();
+    let (promise, deferred) = cx.promise();
+
+    std::thread::spawn(move || {
+        let result = perform_complex_operation();
+    
+        queue.send(move || {
+            let result = cx.number(result);
+
+            deferred.resolve(&cx, result);
+        });
+    });
+
+    Ok(promise)
+}
+```
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Before `Promise`, callbacks of the form `function (err, value)` were very common in JavaScript. Neon has excellent for for these "node style" callbacks in the `Task` trait.
```

Comment:
s/for for/support for/

Reply:
Maybe worth mentioning that the `JsBox` and thread-safe handles RFCs also provide support for asynchronous computations on arbitrary Rust threads, also with classic node-style callbacks.